optuna/integration/skopt.py

```diff
@@ -0,0 +1,219 @@
+
+            If :obj:`None` is specified, :class:`~optuna.samplers.TPESampler` is used
+            as the default. See also :class:`~optuna.samplers`.
+        skopt_kwargs:
+            Keyword arguments passed to the constructor of
+            `skopt.Optimizer <https://scikit-optimize.github.io/#skopt.Optimizer>`_
+            class.
+
+            Note that the ``dimensions`` argument is automatically added by
+            :class:`~optuna.integration.SkoptSampler`, so you don't have to specify it
+            (it will be ignored even if specified).
+
+    """
+
+    def __init__(self, independent_sampler=None, skopt_kwargs=None):
+        # type: (Optional[BaseSampler], Optional[Dict[str, Any]]) -> None
+
+        _check_skopt_availability()
+
+        self._skopt_kwargs = skopt_kwargs or {}
+        if 'dimensions' in self._skopt_kwargs:
+            del self._skopt_kwargs['dimensions']
+
+        self._independent_sampler = independent_sampler or TPESampler()
+
+    def infer_relative_search_space(self, study, trial):
+        # type: (InTrialStudy, FrozenTrial) -> Dict[str, BaseDistribution]
+
+        search_space = {}
+        for name, distribution in study.product_search_space.items():
+            # Skip if the range of the distribution is empty.
+            if isinstance(distribution, distributions.UniformDistribution):
+                if distribution.low == distribution.high:
+                    continue
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                if distribution.low == distribution.high:
+                    continue
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                if distribution.low == distribution.high:
+                    continue
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                if distribution.low == distribution.high:
+                    continue
+
+            search_space[name] = distribution
+
+        return search_space
+
+    def sample_relative(self, study, trial, search_space):
+        # type: (InTrialStudy, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, float]
+
+        if len(search_space) == 0:
+            return {}
+
+        optimizer = _Optimizer(search_space, self._skopt_kwargs)
+        optimizer.tell(study)
+        return optimizer.ask()
+
+    def sample_independent(self, study, trial, param_name, param_distribution):
+        # type: (InTrialStudy, FrozenTrial, str, BaseDistribution) -> float
+
+        return self._independent_sampler.sample_independent(study, trial, param_name,
+                                                            param_distribution)
+
+
+class _Optimizer(object):
+    def __init__(self, search_space, skopt_kwargs=None):
+        # type: (Dict[str, BaseDistribution], Optional[Dict[str, Any]]) -> None
+
+        self._search_space = search_space
+
+        dimensions = []  # type: List[Any]
```

Comment:
I think the type of `dimensions` can be restricted to `List[space.Dimension]`. Do you have any reason to use `List[Any]`?

Reply:
`List[Any]` was necessary in the previous code, but it has become useless, so I removed the type hint completely at the commit https://github.com/pfnet/optuna/pull/399/commits/36b7c7f2a86a5f2095b22c3868a31c7a2f1a2e28.